### PR TITLE
[EVM] Clear the branch condition in TII::removeBranch for stackified code

### PR DESCRIPTION
**Pull request by PavelKopyl**
# Code Review Checklist

## Requirements
- [ ] Have the requirements been met?
- [ ] Have stakeholder(s) approved the change?

## Implementation
- [ ] Does this code change accomplish what it is supposed to do?
- [ ] Can this solution be simplified?
- [ ] Does this change add unwanted compile-time or run-time dependencies?
- [ ] Could an additional framework, API, library, or service improve the solution?
- [ ] Could we reuse part of LLVM instead of implementing the patch or a part of it?
- [ ] Is the code at the right abstraction level?
- [ ] Is the code modular enough?
- [ ] Can a better solution be found in terms of maintainability, readability, performance, or security?
- [ ] Does similar functionality already exist in the codebase? If yes, why isn’t it reused?
- [ ] Are there any best practices, design patterns or language-specific patterns that could substantially improve this code? 

## Logic Errors and Bugs
- [ ] Can you think of any use case in which the
code does not behave as intended?
- [ ] Can you think of any inputs or external events
that could break the code?

## Error Handling and Logging
- [ ] Is error handling done the correct way?
- [ ] Should any logging or debugging information
be added or removed?
- [ ] Are error messages user-friendly?
- [ ] Are there enough log events and are they
written in a way that allows for easy
debugging?

## Maintainability
- [ ] Is the code easy to read?
- [ ] Is the code not repeated (DRY Principle)?
- [ ] Is the code method/class not too long?

## Dependencies
- [ ] Were updates to documentation, configuration, or readme files made as required by this change?
- [ ] Are there any potential impacts on other parts of the system or backward compatibility?

## Security
- [ ] Does the code introduce any security vulnerabilities?

## Performance
- [ ] Do you think this code change decreases
system performance?
- [ ] Do you see any potential to improve the
performance of the code significantly?

## Testing and Testability
- [ ] Is the code testable?
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
	- [ ] For changes to mutable state
- [ ] Do tests reasonably cover the code change (unit/integration/system tests)? 
	- [ ] Line Coverage
	- [ ] Region Coverage
	- [ ] Branch Coverage
- [ ] Are there some test cases, input or edge cases
that should be tested in addition?

## Readability
- [ ] Is the code easy to understand?
- [ ] Which parts were confusing to you and why?
- [ ] Can the readability of the code be improved by
smaller methods?
- [ ] Can the readability of the code be improved by
different function, method or variable names?
- [ ] Is the code located in the right
file/folder/package?
- [ ] Do you think certain methods should be
restructured to have a more intuitive control
flow?
- [ ] Is the data flow understandable?
- [ ] Are there redundant or outdated comments?
- [ ] Could some comments convey the message
better?
- [ ] Would more comments make the code more
understandable?
- [ ] Could some comments be removed by making the code itself more readable?
- [ ] Is there any commented-out code?
- [ ] Have you run a spelling and grammar checker?

## Documentation
- [ ] Is there sufficient documentation?
- [ ] Is the ReadMe.md file up to date?

## Best Practices
- [ ] Follow Single Responsibility principle?
- [ ] Are different errors handled correctly?
- [ ] Are errors and warnings logged?
- [ ] Magic values avoided?
- [ ] No unnecessary comments?
- [ ] Minimal nesting used?

## Experts' Opinion
- [ ] Do you think a specific expert, like a security
expert or a usability expert, should look over
the code before it can be accepted?
- [ ] Will this code change impact different teams, and should they review the change as well?

### CODE REVIEW

File: llvm/lib/Target/EVM/EVMInstrInfo.cpp [hunk 1]
```diff
@@ -187,6 +188,17 @@ unsigned EVMInstrInfo::removeBranch(MachineBasicBlock &MBB,
...
+  const bool IsStackified =
+      MBB.getParent()->getInfo<EVMMachineFunctionInfo>()->getIsStackified();
+  if (IsStackified && HasConditionalBranch) {
+    auto NewMI = BuildMI(&MBB, DebugLoc(), get(EVM::POP_S));
+    // Mark this POP instruction with a special flag to distinguish it from other
+    // POP instruction in insertBranch that is not related to branch-condition
+    // clearing. It’s unlikely that such a case exists; this is mostly for
+    // paranoia check.
+    NewMI->setFlag(MachineInstr::MIFlag::ClearBranchCondition);
+    --Count;
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMInstrInfo.cpp]: Do we need this?
Description of removeBranch says it returns the number of instructions that were removed.

File: llvm/test/CodeGen/EVM/branch-folding-clear-condition.ll [hunk 2]
```diff
@@ -0,0 +1,47 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 2
```
Comment (vladimirradosavljevic) [inline llvm/test/CodeGen/EVM/branch-folding-clear-condition.ll]: I think this should be a MIR test that is calling only branchfolding pass. Could you please convert it?

**Review by vladimirradosavljevic — CHANGES_REQUESTED** [review]
Since we are removing EraVM target, we will remove big chunk of local changes, so I would recommend not to introduce new one, as we have other options (e.g. introduce new `POP_COND_S`).

File: llvm/lib/Target/EVM/EVMInstrInfo.cpp (hunk 1 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMInstrInfo.cpp]: Yeap, but we actually insert 1-byte instruction here, do you meant we shouldn't take it into account?

Comment (vladimirradosavljevic) [reply llvm/lib/Target/EVM/EVMInstrInfo.cpp]: I was looking at callers of `removeBranch` and in some cases they do something based on the return (e.g. `PeelSingleBlockLoop` in `MachineLoopUtils.cpp`). Currently in case we have only conditional branch and we add `POP`, we will return 0, but actually we need to return 1.

**Review by vladimirradosavljevic — APPROVED** [review]
LGTM. Thanks!

File: llvm/lib/Target/EVM/EVMInstrInfo.cpp [hunk 3]
```diff
@@ -187,6 +188,11 @@ unsigned EVMInstrInfo::removeBranch(MachineBasicBlock &MBB,
     MBB.erase(&MBB.back());
     I = MBB.rbegin();
   }
+  const bool IsStackified =
+      MBB.getParent()->getInfo<EVMMachineFunctionInfo>()->getIsStackified();
+  if (IsStackified && HasConditionalBranch)
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMInstrInfo.cpp:225]: I think a part of commit message doc is to be put here.

File: llvm/lib/Target/EVM/EVMInstrInfo.cpp [hunk 4]
```diff
@@ -174,9 +174,10 @@ unsigned EVMInstrInfo::removeBranch(MachineBasicBlock &MBB,
                     << '\n');
 
   unsigned Count = 0;
+  bool HasConditionalBranch = false;
   // Only remove branches from the end of the MBB.
   for (auto I = MBB.rbegin(); I != MBB.rend() && I->isBranch(); ++Count) {
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMInstrInfo.cpp]: getLastNonDebugInstr?

File: llvm/lib/Target/EVM/EVMInstrInfo.cpp [hunk 5]
```diff
@@ -215,6 +221,13 @@ unsigned EVMInstrInfo::insertBranch(MachineBasicBlock &MBB,
     else
       CondOpc = Cond[0].getImm() ? EVM::PseudoJUMPI : EVM::PseudoJUMP_UNLESS;
 
+    // Remove the POP_COND_S instruction that was previously added in
+    // removeBranch to clear the jump condition from the stack.
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMInstrInfo.cpp]: Could you explain why?

**Review by akiramenai — CHANGES_REQUESTED** [review]
Documentation:
1. Could you put some explanations to logic in code?
2. Can we document the rationale for the design choice in commit message?

File: llvm/lib/Target/EVM/EVMInstrInfo.cpp (hunk 5 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMInstrInfo.cpp]: Let's consider basic blocks (BBs) that end with
```
JUMPI
JUMP
```
or
```JUMPI```

When `removeBranch() `is called, it removes all jump instructions.
In stackified code, this can disrupt the stack layout by leaving an unused condition on top of the stack, i.e., condition that would normally be consumed by the `JUMPI` instruction.
To handle this, we add a `POP_COND_S` instruction at the end of the BB after removing the conditional branch, ensuring that the unused condition is properly consumed.
In most cases (I would say in 99%), `removeBranch()` is followed by `insertBranch()`, which typically reinserts a `JUMPI` instruction, possibly with a different target.
In such cases, the condition will again be consumed by the new `JUMPI`, so the `POP_COND_S` is no longer needed and is removed.
Although this adds a small overhead, by creating and then removing `POP_COND_S`, the cost is negligible since this only happens in stackified code.
In the rare cases where ` removeBranch()` is called without a subsequent insertBranch(), the `POP_COND_S` remains and will be emitted as a `POP` instruction.
For example, this happens here:
https://github.com/matter-labs/era-compiler-llvm/blob/8e7767fdb8a19cb687c78827907c10a757a78dc5/llvm/lib/CodeGen/BranchFolding.cpp#L1469

File: llvm/lib/Target/EVM/EVMInstrInfo.cpp (hunk 4 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMInstrInfo.cpp]: We may have a BB ending with
```
JUMPI
JUMP 
```
so with `getLastNonDebugInstr` we will miss the presence of JUMPI.

File: llvm/lib/Target/EVM/EVMInstrInfo.cpp (hunk 3 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMInstrInfo.cpp:225]: OK, added.

**Comment by PavelKopyl** [comment]
> Documentation:
> 
> 1. Could you put some explanations to logic in code?
> 2. Can we document the rationale for the design choice in commit message?

Sure, I've added this to the commit message.

File: llvm/lib/Target/EVM/EVMInstrInfo.cpp (hunk 4 above)
Comment (akiramenai) [reply llvm/lib/Target/EVM/EVMInstrInfo.cpp]: I mean with the loop we could hit a debug instruction and bail out.

**Review by akiramenai — APPROVED** [review]

File: llvm/lib/Target/EVM/EVMInstrInfo.cpp (hunk 4 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMInstrInfo.cpp]: Thanks, got it. I've slightly rewrote the removeBranch to skip debug instructions.

File: llvm/lib/Target/EVM/EVMInstrInfo.cpp [hunk 6]
```diff
@@ -173,20 +173,58 @@ unsigned EVMInstrInfo::removeBranch(MachineBasicBlock &MBB,
...
-  for (auto I = MBB.rbegin(); I != MBB.rend() && I->isBranch(); ++Count) {
-
+  unsigned Count = 0;
+  bool HasConditionalBranch = false;
+  MachineBasicBlock::iterator I = MBB.end();
+  while (I != MBB.begin()) {
+    --I;
+    if (I->isDebugInstr())
+      continue;
+    if (!I->isTerminator())
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMInstrInfo.cpp:184]: This should be `!I->isBranch()`, right?
Terminators are also RET, REVERT, RETURN, STOP and INVALID, and we don't want to remove them.

Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMInstrInfo.cpp:184]: Right, thank you. Pushed a fix: https://github.com/matter-labs/era-compiler-llvm/pull/923
